### PR TITLE
Feature/issue 991 ensure error handler throws to the transit layer

### PR DESCRIPTION
**Pull request by JordanPawlett**
## :memo: Description
Error-handler should re-throw an event error to ensure the messageHandler can return the correct result to the transporter.

### :gem: Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## :checkered_flag: Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] **I have added tests that prove my fix is effective or that my feature works**
- [x] **New and existing unit tests pass locally with my changes**
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by icebob** [comment]
I need to think about it because it can be a breaking change. If somebody throws errors in event handlers in the current version it' catcher and logged. But after this PR, the error will be thrown further and it will cause `UnhandledPromiseRejection` and as of Node 15, it will crashes the process, not just print a warning.

**Comment by JordanPawlett** [comment]
> I need to think about it because it can be a breaking change. If somebody throws errors in event handlers in the current version it' catcher and logged. But after this PR, the error will be thrown further and it will cause `UnhandledPromiseRejection` and as of Node 15, it will crashes the process, not just print a warning.

Ultimately the thrown error is caught and handled in the transit `messageHandler`, and resolved to turn `false`. Does this not negate this problem?

**Comment by icebob** [comment]
> Ultimately the thrown error is caught and handled in the transit `messageHandler`, and resolved to turn `false`. Does this not negate this problem?

Yeah, if the event is received via transported. Locally created events won't be handled.

**Comment by JordanPawlett** [comment]
> > Ultimately the thrown error is caught and handled in the transit `messageHandler`, and resolved to turn `false`. Does this not negate this problem?
> 
> Yeah, if the event is received via transported. Locally created events won't be handled.
> 
> image

Would you be able to point me at the entryPoint for receiving local events rather than via the transporter.
What about a config option to allow the event handler to throw? Or only allow it when `preferLocal` is false?

**Comment by icebob** [comment]
There are multiple places where local events triggered because it depends on the type (emit or broadcast).
Here is the emitted: https://github.com/moleculerjs/moleculer/blob/master/src/service-broker.js#L1366
Here is the broadcasted: https://github.com/moleculerjs/moleculer/blob/master/src/service-broker.js#L1648

File: test/unit/middlewares/error-handler.spec.js [hunk 1]
```diff
@@ -206,5 +203,34 @@
...
+
+				const newHandler = mw.localEvent.call(broker, handler, action);
+				const ctx = Context.create(broker, eventEndpoint);
+				ctx.id = "123456";
+				ctx.nodeID = "server-2";
+
+
+				return newHandler(ctx)
+					.then((res) => {
+						expect(res).toBe(true);
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:224]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
					expect(res).toBe(true);
```

File: test/unit/middlewares/error-handler.spec.js [hunk 2]
```diff
@@ -143,11 +143,11 @@
 			const ctx = Context.create(broker, eventEndpoint);
 
 			return newHandler(ctx)
-				.catch(protectReject)
-				.then(() => {
+				.then(() => { fail('error should be thrown to the transit message handler') })
+				.catch((err) => {
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:147]: Codacy has a fix for the issue: [Replace `(err)` with `err`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
				.catch(err => {
```

File: test/unit/middlewares/error-handler.spec.js [hunk 3]
```diff
@@ -206,5 +203,34 @@
 					expect(broker.logger.error).toHaveBeenCalledWith(err);
 				});
 		});
+
+		describe('if the error handler resolves the error', () => {
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:207]: Codacy has a fix for the issue: [Replace `'if·the·error·handler·resolves·the·error'` with `"if·the·error·handler·resolves·the·error"`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
		describe("if the error handler resolves the error", () => {
```

File: test/unit/middlewares/error-handler.spec.js [hunk 4]
```diff
@@ -143,11 +143,11 @@
 			const ctx = Context.create(broker, eventEndpoint);
 
 			return newHandler(ctx)
-				.catch(protectReject)
-				.then(() => {
+				.then(() => { fail('error should be thrown to the transit message handler') })
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:146]: Codacy has a fix for the issue: [Replace `·fail('error·should·be·thrown·to·the·transit·message·handler')·` with `⏎↹↹↹↹↹fail("error·should·be·thrown·to·the·transit·message·handler");⏎↹↹↹↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
				.then(() => {
```

File: test/unit/middlewares/error-handler.spec.js [hunk 5]
```diff
@@ -206,5 +203,34 @@
...
+
+				return newHandler(ctx)
+					.then((res) => {
+						expect(res).toBe(true);
+						expect(error.ctx).toBe(ctx);
+						expect(broker.errorHandler).toBeCalledTimes(1);
+						expect(broker.errorHandler).toBeCalledWith(error, {
+							ctx,
+							service: action.service,
+							event
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:230]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
						event
```

File: test/unit/middlewares/error-handler.spec.js [hunk 6]
```diff
@@ -206,5 +203,34 @@
...
+				ctx.id = "123456";
+				ctx.nodeID = "server-2";
+
+
+				return newHandler(ctx)
+					.then((res) => {
+						expect(res).toBe(true);
+						expect(error.ctx).toBe(ctx);
+						expect(broker.errorHandler).toBeCalledTimes(1);
+						expect(broker.errorHandler).toBeCalledWith(error, {
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:227]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
					expect(broker.errorHandler).toBeCalledWith(error, {
```

File: test/unit/middlewares/error-handler.spec.js [hunk 7]
```diff
@@ -206,5 +203,34 @@
 					expect(broker.logger.error).toHaveBeenCalledWith(err);
 				});
 		});
+
+		describe('if the error handler resolves the error', () => {
+			beforeEach(() => {
+				broker.errorHandler = jest.fn(() => Promise.resolve(true));
+			});
+
+			it('should return the response from the error handler', () => {
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:212]: Codacy has a fix for the issue: [Replace `'should·return·the·response·from·the·error·handler'` with `"should·return·the·response·from·the·error·handler"`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
			it("should return the response from the error handler", () => {
```

File: test/unit/middlewares/error-handler.spec.js [hunk 8]
```diff
@@ -206,5 +203,34 @@
...
+				ctx.nodeID = "server-2";
+
+
+				return newHandler(ctx)
+					.then((res) => {
+						expect(res).toBe(true);
+						expect(error.ctx).toBe(ctx);
+						expect(broker.errorHandler).toBeCalledTimes(1);
+						expect(broker.errorHandler).toBeCalledWith(error, {
+							ctx,
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:228]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
						ctx,
```

File: test/unit/middlewares/error-handler.spec.js [hunk 9]
```diff
@@ -206,5 +203,34 @@
...
+							service: action.service,
+							event
+						});
+					})
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:232]: Codacy has a fix for the issue: [Replace `↹})` with `});`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
				});
```

File: test/unit/middlewares/error-handler.spec.js [hunk 10]
```diff
@@ -206,5 +203,34 @@
...
+				const ctx = Context.create(broker, eventEndpoint);
+				ctx.id = "123456";
+				ctx.nodeID = "server-2";
+
+
+				return newHandler(ctx)
+					.then((res) => {
+						expect(res).toBe(true);
+						expect(error.ctx).toBe(ctx);
+						expect(broker.errorHandler).toBeCalledTimes(1);
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:226]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
					expect(broker.errorHandler).toBeCalledTimes(1);
```

File: test/unit/middlewares/error-handler.spec.js [hunk 11]
```diff
@@ -206,5 +203,34 @@
...
+						expect(broker.errorHandler).toBeCalledWith(error, {
+							ctx,
+							service: action.service,
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:229]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
						service: action.service,
```

File: test/unit/middlewares/error-handler.spec.js [hunk 12]
```diff
@@ -206,5 +203,34 @@
...
+							event
+						});
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:231]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
					});
```

File: test/unit/middlewares/error-handler.spec.js [hunk 13]
```diff
@@ -206,5 +203,34 @@
...
+				const newHandler = mw.localEvent.call(broker, handler, action);
+				const ctx = Context.create(broker, eventEndpoint);
+				ctx.id = "123456";
+				ctx.nodeID = "server-2";
+
+
+				return newHandler(ctx)
+					.then((res) => {
+						expect(res).toBe(true);
+						expect(error.ctx).toBe(ctx);
```
Comment (icebob) [inline test/unit/middlewares/error-handler.spec.js:225]: Codacy has a fix for the issue: [Delete `↹`](https://app.codacy.com/gh/moleculerjs/moleculer/pullRequest?prid=8141835)
```suggestion
					expect(error.ctx).toBe(ctx);
```

**Pull request closed without merging**